test_conformance/extensions/cl_khr_command_buffer/command_buffer_pipelined_enqueue.cpp

```diff
@@ -0,0 +1,318 @@
+        test_error(error, "clEnqueueCommandBufferKHR failed");
+
+        std::vector<cl_int> output_data(num_elements);
+        error = clEnqueueReadBuffer(out_of_order_queue, out_mem, CL_TRUE, 0,
+                                    data_size(), output_data.data(), 1,
+                                    &events[2], nullptr);
+        test_error(error, "clEnqueueReadBuffer failed");
+
+        // Verify
+        const cl_int ref = pattern + 1;
+        for (size_t i = 0; i < num_elements; i++)
+        {
+            CHECK_VERIFICATION_ERROR(ref, output_data[i], i);
+        }
+        return CL_SUCCESS;
+    }
+
+    clCommandQueueWrapper out_of_order_queue;
+    clCommandBufferWrapper out_of_order_command_buffer;
+    clEventWrapper events[3] = { nullptr, nullptr, nullptr };
+};
+
+struct BarrierPipelined : public CommandBufferPipelined
+{
+    BarrierPipelined(cl_device_id device, cl_context context,
+                     cl_command_queue queue)
+        : CommandBufferPipelined(device, context, queue),
+          out_of_order_queue(nullptr), out_of_order_command_buffer(this)
+    {}
+
+    bool Skip() override
+    {
+        return CommandBufferPipelined::Skip() || !out_of_order_support;
+    }
+
+    cl_int SetUp(int elements) override
+    {
+        cl_int error = CommandBufferPipelined::SetUp(elements);
+        test_error(error, "EventPipelined::SetUp failed");
+
+        out_of_order_queue = clCreateCommandQueue(
+            context, device, CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE, &error);
+        test_error(error, "Unable to create command queue to test with");
+
+        out_of_order_command_buffer =
+            clCreateCommandBufferKHR(1, &out_of_order_queue, nullptr, &error);
+        test_error(error, "clCreateCommandBufferKHR failed");
+
+        return CL_SUCCESS;
+    }
+
+    cl_int Run() override
+    {
+        cl_int error = RecordCommandBuffer(out_of_order_command_buffer);
+        test_error(error, "RecordCommandBuffer failed");
+
+        error = clEnqueueCommandBufferKHR(
+            0, nullptr, out_of_order_command_buffer, 0, nullptr, nullptr);
+        test_error(error, "clEnqueueCommandBufferKHR failed");
+
+        error = clEnqueueBarrier(out_of_order_queue);
+        test_error(error, "clEnqueueBarrier failed");
+
+        error =
+            clEnqueueNDRangeKernel(out_of_order_queue, mul_kernel, 1, nullptr,
+                                   &num_elements, nullptr, 0, nullptr, nullptr);
+        test_error(error, "clEnqueueNDRangeKernel failed");
+
+        error = clEnqueueBarrier(out_of_order_queue);
+        test_error(error, "clEnqueueBarrier failed");
+
+        error = clEnqueueCommandBufferKHR(
+            0, nullptr, out_of_order_command_buffer, 0, nullptr, nullptr);
+        test_error(error, "clEnqueueCommandBufferKHR failed");
+
+        std::vector<cl_int> output_data(num_elements);
+        error = clEnqueueReadBuffer(out_of_order_queue, out_mem, CL_TRUE, 0,
```

Comment:
Hello Ewan, sorry for not catching this earlier, but I believe there needs to be a barrier before the clEnqueueReadBuffer, as there is no guarantee it will wait until all work is done (out of order queue).